File: client/lib/core/download/download_state.dart
enum DownloadStatus { idle, downloading, extracting, complete, failed }

class DownloadState {
  final DownloadStatus status;
  final double progress;
  final String? localPath;
  final String? extractedDir;
  final String? error;

  const DownloadState({
    this.status = DownloadStatus.idle,
    this.progress = 0,
    this.localPath = null,
    this.extractedDir = null,
    this.error = null,
  });

  const DownloadState.idle() : this();

  bool get isDownloaded => status == DownloadStatus.complete && localPath != null;
  bool get isExtracted => extractedDir != null;
  bool get isAvailableOffline => isDownloaded;

  DownloadState copyWith({
    DownloadStatus? status,
    double? progress,
    String? localPath,
    String? extractedDir,
    String? error,
  }) =>
      DownloadState(
        status: status ?? this.status,
        progress: progress ?? this.progress,
        localPath: localPath ?? this.localPath,
        extractedDir: extractedDir ?? this.extractedDir,
        error: error ?? this.error,
      );
}

// ---------------------------------------------------------------------------
// Folder-level download state
// ---------------------------------------------------------------------------

enum FolderDownloadStatus { idle, fetching, downloading, complete, failed }

class FolderDownloadState {
  final FolderDownloadStatus status;
  final int total;
  final int completed;
  final String? error;

  const FolderDownloadState({
    this.status = FolderDownloadStatus.idle,
    this.total = 0,
    this.completed = 0,
    this.error,
  });

  const FolderDownloadState.idle()
      : status = FolderDownloadStatus.idle,
        total = 0,
        completed = 0,
        error = null;

  double get progress => total == 0 ? 0 : completed / total;
}

File: android/app/src/main/java/com/rekindle/app/core/download/DownloadState.kt
package com.rekindle.app.core.download

enum class DownloadStatus { IDLE, DOWNLOADING, EXTRACTING, COMPLETE, FAILED }

data class DownloadState(
    val status: DownloadStatus = DownloadStatus.IDLE,
    val progress: Float = 0f,
    val localPath: String? = null,
    val extractedDir: String? = null,
    val error: String? = null,
) {
    val isDownloaded: Boolean get() = status == DownloadStatus.COMPLETE && localPath != null
    val isAvailableOffline: Boolean get() = isDownloaded
    val isExtracted: Boolean get() = extractedDir != null
}

// ---------------------------------------------------------------------------
// Folder-level download state
// ---------------------------------------------------------------------------

enum class FolderDownloadStatus { IDLE, FETCHING, DOWNLOADING, COMPLETE, FAILED }

data class FolderDownloadState(
    val status: FolderDownloadStatus = FolderDownloadStatus.IDLE,
    val total: Int = 0,
    val completed: Int = 0,
    val error: String? = null,
) {
    val progress: Float get() = if (total == 0) 0f else completed.toFloat() / total
}
